src/udev/dmi_memory_id/dmi_memory_id.c

```diff
@@ -399,7 +399,7 @@ static void dmi_memory_device_manufacturer_id(
         /* LSB is 7-bit Odd Parity number of continuation codes */
         if (code != 0)
                 printf("MEMORY_DEVICE_%u_%s=Bank %d, Hex 0x%02X\n", slot_num, attr_suffix,
-                       (code & 0x7F) + 1, code >> 8);
+                       (code & 0x7F) + 1, (unsigned) (code >> 8));
```

Comment:
maybe change the `%d` to `%"PRIu16"` and the `0x%02X\n` to `0x%02"PRIx16"\n"`

or so?

Reply:
How is this adjusted? You still get integer promotion from `code >> 8` to `int`

Reply:
nope, you don't afaiu.

`code` is going to be promoted from uint16_t to `unsigned int` iiuc. The following `>> 8` does not change the type, so it's still `unsigned int`. PRIx16 is the right format string for that, since it already takes expected integer propagation into account!

Hence, afaics just use PRIx16 and things should be fine.

Reply:
@poettering 

```
        uint16_t code = 16;
        static_assert(__builtin_types_compatible_p(typeof(code << 8), unsigned), "is not unsigned");
```

This fails to compile with:

```
../src/pcrlock/pcrlock-firmware.c:130:9: error: static assertion failed: "is not unsigned"
  130 |         static_assert(__builtin_types_compatible_p(typeof(code << 8), unsigned), "is not unsigned");
```

Changing to `int` makes it compile.

Need more evidence??

Reply:
I don't understand integer promotion rules then. Makes no sense to me. somebody please explain!

Reply:
Anything smaller than `int` gets promoted to `int` unless it doesn't fit in `int` then it gets promoted to `unsigned int`. Completely stupid but not that hard to grok no?

Reply:
so, you are saying they change signedness? that's just weird.

Reply:
or is the signedness of the result dependent on the signedness of the rhs of the >>? that'd be wuite weird, but does that work, if you change 8 → 8U?

If not, I guess the cast is kinda necessary, but it would be more logic to use `(uint16_t)` rather than `(unsigned)`, no? We are not increasing the width here really after all.

Reply:
I double checked, `8U` does not work.